substrate/frame/treasury/src/migration.rs

```diff
@@ -0,0 +1,74 @@
+use super::*;
+use alloc::collections::BTreeSet;
+use core::marker::PhantomData;
+use frame_support::traits::UncheckedOnRuntimeUpgrade;
+
+/// The log target for this pallet.
+const LOG_TARGET: &str = "runtime::treasury";
+
+mod v1 {
+	use super::*;
+	pub struct MigrateToV1Impl<T, I>(PhantomData<(T, I)>);
+
+	impl<T: Config<I>, I: 'static> UncheckedOnRuntimeUpgrade for MigrateToV1Impl<T, I> {
+		fn on_runtime_upgrade() -> frame_support::weights::Weight {
+			let mut approval_index = BTreeSet::new();
+			for approval in Approvals::<T, I>::get().iter() {
+				approval_index.insert(*approval);
+			}
+
+			let mut proposals_released = 0;
+			for (proposal_index, p) in Proposals::<T, I>::iter() {
+				if !approval_index.contains(&proposal_index) {
+					let err_amount = T::Currency::unreserve(&p.proposer, p.bond);
+					debug_assert!(err_amount.is_zero());
+					Proposals::<T, I>::remove(proposal_index);
+					log::info!(
+						target: LOG_TARGET,
+						"Released bond amount of {:?} to proposer {:?}",
+						p.bond,
+						p.proposer,
+					);
+					proposals_released += 1;
+				}
+			}
+
+			log::info!(
+				target: LOG_TARGET,
+				"Storage migration v1 for pallet-treasury finished, released {} proposal bonds.",
+				proposals_released,
+			);
+
+			// calculate and return migration weights
+			let approvals_read = 1;
+			T::DbWeight::get()
+				.reads_writes(proposals_released as u64 + approvals_read, proposals_released as u64)
```

Comment:
we also need to account for the weight of `Currency::unreserve`. we do not know really it's weight here. we can specify type parameter that implements `Get<Weight>` that would return a weight for a single unreserve.

Reply:
Added `<T as pallet_balances::Config>::WeightInfo::force_unreserve() * proposals_released` weight, I think it should be appropriate value because treasury pallet assumes knowledge about balances pallet used?